gotham/src/service/mod.rs

```diff
@@ -0,0 +1,168 @@
+//! Defines the `Service` which is used by a Gotham application to interface to Hyper.
+
+use std::{io, thread};
+use std::net::SocketAddr;
+use std::sync::Arc;
+use std::panic::AssertUnwindSafe;
+
+use hyper;
+use hyper::server::{NewService, Service};
+use hyper::{Request, Response};
+use futures::Future;
+use tokio_core::reactor::Handle;
+
+use handler::NewHandler;
+use state::{request_id, set_request_id, State};
+use state::client_addr::put_client_addr;
+use http::request::path::RequestPathSegments;
+
+mod timing;
+mod trap;
+
+/// Wraps a `NewHandler` to provide a `hyper::server::NewService` implementation for Gotham
+/// handlers.
+pub(super) struct GothamService<T>
+where
+    T: NewHandler + 'static,
+{
+    t: Arc<T>,
+    handle: Handle,
+}
+
+impl<T> GothamService<T>
+where
+    T: NewHandler + 'static,
+{
+    /// Creates a `GothamService` for the given `NewHandler`.
+    /// # }
+    /// ```
+    pub(super) fn new(t: Arc<T>, handle: Handle) -> GothamService<T> {
+        GothamService { t, handle }
+    }
+
+    pub(super) fn connect(&self, client_addr: SocketAddr) -> ConnectedGothamService<T> {
+        ConnectedGothamService {
+            t: self.t.clone(),
+            handle: self.handle.clone(),
+            client_addr,
+        }
+    }
+}
+
+pub(super) struct ConnectedGothamService<T>
+where
+    T: NewHandler + 'static,
+{
+    t: Arc<T>,
+    handle: Handle,
+    client_addr: SocketAddr,
+}
+
+impl<T> Clone for ConnectedGothamService<T>
+where
+    T: NewHandler + 'static,
+{
+    fn clone(&self) -> Self {
+        ConnectedGothamService {
+            t: self.t.clone(),
+            handle: self.handle.clone(),
+            client_addr: self.client_addr,
+        }
+    }
+}
+
+impl<T> NewService for ConnectedGothamService<T>
```

Comment:
I've looked into the types and I believe I've convinced myself that `ConnectedGothamService` does not actually need to be `NewService` any longer, which is turn simplifies code in `clone_sockets.rs` and various test cases.

Is there any benefit to this remaining in place I may not be considering?

Reply:
Example of the code I hacked on with `NewService` removed if useful:

``` rust
    core.run(listener.incoming().for_each(|(socket, addr)| {
        let service = new_service.connect(addr);
        let f = protocol.serve_connection(socket, service).then(|_| Ok(()));
        //.map(|_| ())
        //.map_err(|_| ());

        handle.spawn(f);
        Ok(())
    })).expect("unable to run reactor over listener");
```